### DISCUSSION TITLE
How to access `setNodes` inside a custom node/edge?

### DISCUSSION THREAD

**Discussion by edhyah** [xyflow/xyflow#2514] [Q&A]
Hello, probably a simple question, but can't find the answer anywhere.

How do I pass in `setNodes` into the `data` field for custom edges/nodes?

In other words, I have a custom edge with a button that hides a target node. How can I call `setNodes` in this custom node to do so? I tried the following
```
const reactFlowInstance = useReactFlow()
...
reactFlowInstance.setNodes((nodes) => {
  nodes.map((node) => {
    if (node.id === nodeId) {
      node.hidden = !node.hidden;
    }
    return node;
  });
});
```
Doing the above runs into an error saying that `nextNodes` is null.

I also tried using `useStoreApi` but couldn't get that to work.

Is there another way I'm missing?

**Comment by edhyah** [comment]
Ah, sorry just found this link: https://reactflow.dev/docs/guides/state-management/

I'll use that.

Either way, should I avoid using `getReactFlow()` and `getStoreApi()` then to update the state from a custom node?

**Reply by moklick** [reply]
You can use a state management library, a React context where you store the `setNodes` function of your local state, or `const { setNodes } = useReactFlow()`.

**Comment by yhojann-cl** [comment]
You need return the array from `nodes.map` to `setNodes`, in your code call the map but can not return the result, in the official website the array is returned:

https://reactflow.dev/examples/nodes/update-node

```javascript
    setNodes((nds) =>
      nds.map((node) => {
```

But in your code the function is enclosed without return the value:

```javascript
reactFlowInstance.setNodes((nodes) => {
  nodes.map((node) => {
```

Add a `return` statement or remove keys:

```javascript
reactFlowInstance.setNodes((nodes) => {
  return nodes.map((node) => {
```

```javascript
reactFlowInstance.setNodes((nodes) =>
  nodes.map((node) => {
```
